File: backend/src/processing/simple_task.rs
/// Evelyn: Your personal assistant, project manager and calendar
/// Copyright (C) 2017 Gregory Jensen
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::sync::Arc;

use serde_json;

use server::routing::{RouterInput, RouterOutput};
use model;
use processing;
use core::simple_task;
use core::error_messages::EvelynServiceError;


pub fn create_simple_task_processor(router_input: RouterInput, processor_data: Arc<processing::ProcessorData>) -> RouterOutput {
    let request_model_de: Result<model::simple_task::CreateSimpleTaskModel,_> = serde_json::from_str(&router_input.request_body);

    match request_model_de {
      Ok(request_model) => {
        match simple_task::create_simple_task(request_model, processor_data) {
            Ok(response) => {
                RouterOutput{response_body: serde_json::to_string(&response).unwrap()}
            }
            Err(e) => {
                RouterOutput{
                    response_body: serde_json::to_string(&model::simple_task::CreateSimpleTaskResponseModel {
                        tasks: None,
                        error: Some(From::from(EvelynServiceError::FailedToCreateSimpleTask(e))),
                    }).unwrap()
                }
            }
        }
      },
      Err(e) => {
        let response = model::simple_task::LookupSimpleTaskResponseModel {
            tasks: None,
            error: Some(From::from(EvelynServiceError::CouldNotDecodeTheRequestPayload(e)))
        };

        RouterOutput{response_body: serde_json::to_string(&response).unwrap()}
      }
    }
}

pub fn lookup_simple_task_processor(router_input: RouterInput, processor_data: Arc<processing::ProcessorData>) -> RouterOutput {
    let request_model_de: Result<model::simple_task::LookupSimpleTaskRequestModel,_> = serde_json::from_str(&router_input.request_body);

    match request_model_de {
      Ok(request_model) => {
        match simple_task::lookup_simple_tasks(request_model, processor_data) {
            Ok(response) => {
                RouterOutput{response_body: serde_json::to_string(&response).unwrap()}
            }
            Err(e) => {
                RouterOutput{
                    response_body: serde_json::to_string(&model::simple_task::LookupSimpleTaskResponseModel {
                        tasks: None,
                        error: Some(From::from(EvelynServiceError::FailedToLookupSimpleTask(e))),
                    }).unwrap()
                }
            }
        }
      },
      Err(e) => {
        let response = model::simple_task::LookupSimpleTaskResponseModel {
            tasks: None,
            error: Some(From::from(EvelynServiceError::CouldNotDecodeTheRequestPayload(e)))
        };

        RouterOutput{response_body: serde_json::to_string(&response).unwrap()}
      }
    }
}

pub fn update_simple_task_processor(router_input: RouterInput, processor_data: Arc<processing::ProcessorData>) -> RouterOutput {
    let request_model_de: Result<model::simple_task::UpdateSimpleTaskRequestModel,_> = serde_json::from_str(&router_input.request_body);

    match request_model_de {
      Ok(request_model) => {
        match simple_task::update_simple_task(request_model, processor_data) {
            None => {
                RouterOutput{response_body: serde_json::to_string(&model::simple_task::UpdateSimpleTaskResponseModel {
                    error: None,
                }).unwrap()}
            },
            Some(e) => {
                let model: model::ErrorModel = From::from(EvelynServiceError::FailedToUpdateSimpleTask(e));
                RouterOutput {
                    response_body: serde_json::to_string(&model::simple_task::UpdateSimpleTaskResponseModel {
                        error: Some(model),
                    }).unwrap()
                }
            }
        }
      },
      Err(e) => {
          let model: model::ErrorModel = From::from(EvelynServiceError::CouldNotDecodeTheRequestPayload(e));
          RouterOutput {
              response_body: serde_json::to_string(&model::user::CreateUserResponseModel {
                  error: Some(model),
              }).unwrap()
          }
      }
    }
}

File: test/server_integration_tests/spec/simple_task.js
// Evelyn: Your personal assistant, project manager and calendar
// Copyright (C) 2017 Gregory Jensen
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

if (!global.Promise) {
    global.Promise = require('bluebird');
}

var chai = require('chai');
var chaiSubset = require('chai-subset');
var moment = require('moment');

var httpHelper = require('../helpers/chai_http_request_helper.js');
var commonRequestsHelper = require('../helpers/common_requests_helper.js');

var simpleTaskHelper = require('../helpers/spec_helpers/simple_task_helper.js');

chai.use(chaiSubset);
var expect = chai.expect;

describe('Simple Task', function() {
    var date = new Date ().toISOString();
    var token = null;

    before(function () {
        return commonRequestsHelper.chaiHttpPostPurgeDatabase()
        .then(function () {
            return commonRequestsHelper.createUserAndLogon();
        })
        .then(function (_token) {
            token = _token;
        });
    });

    it('Creates a task', function() {
        return httpHelper.post(
            '/simpletask/create',
            {
                Token: token,
                Title: "Test Task",
                Description: "Descriptive",
                DueDate: date
            }
        )
        .then(function (response) {
            expect(response.Error).to.be.null;
            expect(response.TaskId).to.be.okay;
        });
    });

    describe("Lookup", function () {
        before(function () {
            var task_id;
            var simpletask = {
                Token : token,
                Title : "Test Task",
                Description : "Descriptive",
                DueDate : date
            };

            // Lookup fetches multiple tasks, create multiple tasks to test
            return commonRequestsHelper.chaiHttpPostPurgeDatabase()
            .then(function (response) {
                return simpleTaskHelper.createTasks(simpletask, 12)
            })
            .then(function (response) {
                expect(response.Error).to.be.null;

                return simpleTaskHelper.lookupTasks(token);
            })
            .then(function (response) {
                expect(response.Error).to.be.null;

                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.lengthOf(12);
                expect(response.SimpleTasks[0]).to.have.property('TaskId').that.is.a('string');

                return httpHelper.post(
                    '/simpletask/update',
                    {
                        Token: token,
                        TaskId: response.SimpleTasks[0].TaskId,
                        NewTitle: simpletask.Title,
                        NewDescription: simpletask.Description,
                        NewDueDate: simpletask.DueDate,
                        NewCompleted: true
                    }
                );
            })
            .then(function (response) {
                expect(response.Error).to.be.null;
            });
        });

        it('Fetches unlimited unfinished tasks', function() {
            return httpHelper.post(
                '/simpletask/lookup',
                {
                    Token: token,
                    ShowCompleted: false,
                    Limit: 0
                }
            )
            .then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.not.containSubset([{Completed: true}]);
            });
        });

        it('Fetches 10 unfinished tasks', function() {
            return httpHelper.post(
                '/simpletask/lookup',
                {
                    Token: token,
                    ShowCompleted: false,
                    Limit: 10
                }
            )
            .then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.have.length.of.at.most(10);
                expect(response.SimpleTasks).to.not.containSubset([{Completed: true}]);
            });
        });

        it('Fetches unlimited tasks, including completed', function() {
            return httpHelper.post(
                '/simpletask/lookup',
                {
                    Token: token,
                    ShowCompleted: true,
                    Limit: 0
                }
            )
            .then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.containSubset([{Completed: true}]);
            });
        });

        it('Fetches 10 tasks, including completed', function() {
            return httpHelper.post(
                '/simpletask/lookup',
                {
                    Token: token,
                    ShowCompleted: true,
                    Limit: 10
                }
            )
            .then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.have.length.of.at.most(10);
                expect(response.SimpleTasks).to.containSubset([{Completed: true}]);
            });
        });
    });

    describe("Update", function () {
        var simpletask = {};

        before(function () {
            simpletask.title = "Test Task";
            simpletask.description = "Descriptive";
            simpletask.dueDate = date;
            simpletask.completed = false;

            return httpHelper.post(
                '/simpletask/create',
                {
                    Token : token,
                    Title : simpletask.title,
                    Description : simpletask.description,
                    DueDate : simpletask.dueDate,
                }
            )
            .then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.TaskId).to.not.be.null;
                simpletask.taskId = response.TaskId;
            });
        });

        it('Changes title', function() {
            var newTitle = "This is a new title";
            return httpHelper.post(
                '/simpletask/update',
                {
                    Token: token,
                    TaskId: simpletask.taskId,
                    NewTitle: newTitle,
                    NewDescription: simpletask.description,
                    NewDueDate: simpletask.dueDate,
                    NewCompleted: simpletask.completed
                }
            ).then(
                function (response) {
                    expect(response.Error).to.be.null;
                    return httpHelper.post(
                        '/simpletask/lookup',
                        {
                            Token : token,
                            Limit : 0,
                            ShowCompleted : true
                        }
                    );
                }
            ).then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.containSubset([{TaskId: simpletask.taskId, Title: newTitle}]);
            });
        });

        it('Changes description', function() {
            var newDescription = "New Description";
            return httpHelper.post(
                '/simpletask/update',
                {
                    Token: token,
                    TaskId: simpletask.taskId,
                    NewTitle: simpletask.title,
                    NewDescription: newDescription,
                    NewDueDate: simpletask.dueDate,
                    NewCompleted: simpletask.completed,
                }
            ).then(
                function (response) {
                    expect(response.Error).to.be.null;
                    return httpHelper.post(
                        '/simpletask/lookup',
                        {
                            Token : token,
                            Limit : 0,
                            ShowCompleted : true
                        }
                    );
                }
            ).then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.containSubset([{TaskId: simpletask.taskId, Description: newDescription}]);
            });
        });

        it('Changes duedate', function() {
            var newDate = moment().add(3, 'days').toISOString();

            return httpHelper.post(
                '/simpletask/update',
                {
                    Token: token,
                    TaskId: simpletask.taskId,
                    NewTitle: simpletask.title,
                    NewDescription: simpletask.description,
                    NewDueDate: newDate,
                    NewCompleted: simpletask.completed,
                }
            ).then(function (response) {
                expect(response.Error).to.be.null;
                return httpHelper.post(
                    '/simpletask/lookup',
                    {
                        Token : token,
                        Limit : 0,
                        ShowCompleted : true
                    }
                );
            })
            .then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.containSubset([{TaskId: simpletask.taskId, DueDate: newDate}]);
            });
        });

        it('Mark as complete', function() {
            return httpHelper.post(
                '/simpletask/update',
                {
                    Token: token,
                    TaskId: simpletask.taskId,
                    NewTitle: simpletask.title,
                    NewDescription: simpletask.description,
                    NewDueDate: simpletask.dueDate,
                    NewCompleted: true,
                }
            ).then(
                function (response) {
                    expect(response.Error).to.be.null;
                    return httpHelper.post(
                        '/simpletask/lookup',
                        {
                            Token : token,
                            Limit : 0,
                            ShowCompleted : true
                        }
                    );
                }
            ).then(function (response) {
                expect(response.Error).to.be.null;
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.length.of.at.least(1);
                expect(response.SimpleTasks).to.containSubset([{TaskId: simpletask.taskId, Completed: true}]);
            });
        });
    });

    describe('Remove', function() {
        before(function() {
            return commonRequestsHelper.chaiHttpPostPurgeDatabaseArea('simpletask');
        });

        it('Removes a task', function() {
            var taskIdToRemove = null;

            return simpleTaskHelper.createTasks({
                Token : token,
                Title : "Test Task",
                Description : "Descriptive",
                DueDate : date
            }, 2)
            .then(function() {
                return simpleTaskHelper.lookupTasks(token);
            })
            .then(function(response) {
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.lengthOf(2);

                taskIdToRemove = response.SimpleTasks[0].taskId;
                expect(response.SimpleTasks[1].TaskId).to.not.equal(taskIdToRemove);

                return httpHelper.post('/simpletask/remove', {
                    Token: token,
                    TaskId: response.SimpleTasks[0].TaskId
                });
            })
            .then(function(response) {
                expect(response.Error).to.be.null;

                return simpleTaskHelper.lookupTasks(token);
            })
            .then(function(response) {
                expect(response.SimpleTasks).to.be.an.array;
                expect(response.SimpleTasks).to.have.lengthOf(1);
                expect(response.SimpleTasks[0].TaskId).to.not.equal(taskIdToRemove);
            });
        });
    });
});
